From 9c777a73b5fb5da6a155ba13a58389122287b07f Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Fri, 1 Jul 2016 10:39:47 +0300
Subject: [PATCH] Don't try to access undefined app list of AppsServiceChild.
 Fixes MER#1609

When accessing e.g. imgur.com for the first time list build up fails
(.mozilla/mozembed is still clean). Subsequental access are ok.

This backout partically changes from upstream commit sha1 55152ad9d9a23874

Signed-off-by: Raine Makelainen <raine.makelainen@jolla.com>
---
 dom/apps/AppsServiceChild.jsm | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/dom/apps/AppsServiceChild.jsm b/dom/apps/AppsServiceChild.jsm
index 4aca938af4dd..f8cfd87ad0ae 100644
--- a/dom/apps/AppsServiceChild.jsm
+++ b/dom/apps/AppsServiceChild.jsm
@@ -103,16 +103,19 @@ this.DOMApplicationRegistry = {
       this.cpmm.addMessageListener(aMsgName, this);
     }).bind(this));
 
-    this.webapps = { };
-    // We need a fast mapping from localId -> app, so we add an index.
-    // We also add the manifest to the app object.
-    this.localIdIndex = { };
-    for (let id in this.webapps) {
-      let app = this.webapps[id];
-      this.localIdIndex[app.localId] = app;
-      app.manifest = list.manifests[id];
+    try {
+      this.webapps = { };
+      // We need a fast mapping from localId -> app, so we add an index.
+      // We also add the manifest to the app object.
+      this.localIdIndex = { };
+      for (let id in this.webapps) {
+        let app = this.webapps[id];
+        this.localIdIndex[app.localId] = app;
+        app.manifest = list.manifests[id];
+      }
+    } catch (e) {
+      debug("Error while building fast maps: " + e + "\n");
     }
-
     Services.obs.addObserver(this, "xpcom-shutdown", false);
   },
 
-- 
2.17.1

